Mathlib/Analysis/NormedSpace/Star/ContinuousFunctionalCalculus/Restrict.lean

```diff
@@ -0,0 +1,117 @@
+
+/-! # Restriction of the continuous functional calculus to a scalar subring
+
+The main declaration in this file is:
+
++ `SpectrumRestricts.cfc`: builds a continuous functional calculus over a subring of scalars.
+  This is use for automatically deriving the continuous functional calculi on selfadjoint or
+  positive elements from the one for normal elements.
+
+This will allow us to take an instance of the
+`ContinuousFunctionalCalculus ℂ IsStarNormal` and produce both of the instances
+`ContinuousFunctionalCalculus ℝ (IsSelfAdjoint` and `ContinuousFunctionalCalculus ℝ≥0 (0 ≤ ·)`
+simply by proving:
+
+1. `IsSelfAdjoint x ↔ IsStarNormal x ∧ SpectrumRestricts Complex.re x`,
+2. `0 ≤ x ↔ IsSelfAdjoint x ∧ SpectrumRestricts Real.toNNReal x`.
+-/
+
+namespace SpectrumRestricts
+
+lemma compactSpace {R S A : Type*} [CommSemiring R] [CommSemiring S] [Ring A]
+    [Algebra R S] [Algebra R A] [Algebra S A] [IsScalarTower R S A] [TopologicalSpace R]
+    [TopologicalSpace S] {a : A} (f : C(S, R)) (h : SpectrumRestricts a f)
+    [h_cpct : CompactSpace (spectrum S a)] : CompactSpace (spectrum R a) := by
+  rw [← isCompact_iff_compactSpace] at h_cpct ⊢
+  exact h.image ▸ h_cpct.image (map_continuous f)
+
+universe u v w
+
+/-- If the spectrum of an element restricts to a smaller scalar ring, then a continuous functional
+calculus over the larger scalar ring descends to the smaller one. -/
+@[simps!]
+def starAlgHom {R : Type u} {S : Type v} {A : Type w} [CommSemiring R]
+    [StarRing R] [TopologicalSpace R] [TopologicalSemiring R] [ContinuousStar R] [CommSemiring S]
+    [StarRing S] [TopologicalSpace S] [TopologicalSemiring S] [ContinuousStar S] [Ring A]
+    [StarRing A] [Algebra R S] [Algebra R A] [Algebra S A]
+    [IsScalarTower R S A] [StarModule R S] [ContinuousSMul R S] {a : A}
+    (φ : C(spectrum S a, S) →⋆ₐ[S] A) (f : C(S, R)) (h : SpectrumRestricts a f) :
+    C(spectrum R a, R) →⋆ₐ[R] A :=
+  (φ.restrictScalars R).comp <|
+    (ContinuousMap.compStarAlgHom (spectrum S a) (.ofId R S) (algebraMapCLM R S).continuous).comp <|
+      ContinuousMap.compStarAlgHom' R R
+        ⟨Subtype.map f h.subset_preimage, (map_continuous f).subtype_map
+          fun x (hx : x ∈ spectrum S a) => h.subset_preimage hx⟩
+
+variable {R S A : Type*} {p q : A → Prop}
+variable [CommSemiring R] [StarRing R] [MetricSpace R] [TopologicalSemiring R] [ContinuousStar R]
+variable [CommSemiring S] [StarRing S] [MetricSpace S] [TopologicalSemiring S] [ContinuousStar S]
+variable [TopologicalSpace A] [Ring A] [StarRing A] [Algebra S A] [ContinuousFunctionalCalculus S q]
+variable [Algebra R S] [Algebra R A] [IsScalarTower R S A] [StarModule R S] [ContinuousSMul R S]
+
+/-- Given a `ContinuousFunctionalCalculus S q`. If we form the predicate `p` for characterized by
+`q a` and the spectrum of `a` restricts to the scalar subring `R` via `f : C(S, R)`, then we can
+get a restricted functional calculus `ContinuousFunctionalCalculus R p`. -/
```

Comment:
I can't parse the "for characterized" bit in that docstring.

Reply:
better?

Reply:
Yes!